tech_reports/LLMs/llms.md

```diff
@@ -81,10 +81,26 @@ Authors:
+
+#### Accuracy Testing
+
+In order to thoroughly test the accuracy of a model, a bottom up approach is taken such that sub-modules of the model are tested all the way up to the full token generation.
+- **Sub-module unit tests**: Each sub-module of the model should have its own test. For example, the [llama3 models](https://github.com/tenstorrent/tt-metal/tree/main/models/demos/llama3) have a separate [MLP test](https://github.com/tenstorrent/tt-metal/blob/main/models/demos/llama3/tests/test_llama_mlp.py), [attention test](https://github.com/tenstorrent/tt-metal/blob/main/models/demos/llama3/tests/test_llama_attention.py), and [decoder layer test](https://github.com/tenstorrent/tt-metal/blob/main/models/demos/llama3/tests/test_llama_decoder.py). For each of these tests, the outputs produced by the TT implementation of the model are compared against those of the original reference model (typically from Hugging Face) on CPU for a small set of inputs. The metric used for these unit tests is typically the [Pearson Correlation Coefficient (PCC)](https://en.wikipedia.org/wiki/Pearson_correlation_coefficient), where a PCC of 1 indicates a perfect positive correlation between the two sets of outputs, and a PCC of 0 indicates no linear correlation. A good rule of thumb is that the MLP, attention, and other small sub-modules should have a PCC of ~0.999, while a PCC of ~0.998 would be reasonable for a full decoder layer.
+- **Model-level unit tests**: In addition to the sub-module unit tests, there should also be unit tests for a full layer of the model with all sub-modules, and the full model comprising of all layers. For example, the [llama3 model test](https://github.com/tenstorrent/tt-metal/blob/main/models/demos/llama3/tests/test_llama_model.py) runs 1 or many layers of the model over multiple iterations and checks the PCC against the reference model. A rule of thumb is that the full model PCC should be approximately ~0.99.
+- **Dataset evaluation**: Once a model has been brought up with sufficient accuracy on the smaller unit tests, it should be tested on a larger set of prompts such as a full dataset or a subset of it. For example, the [Falcon7b perplexity test](https://github.com/tenstorrent/tt-metal/blob/main/models/demos/falcon7b_common/tests/perplexity/test_perplexity_falcon.py) loads a subset of the [WikiText dataset](https://huggingface.co/datasets/Salesforce/wikitext) and computes several metrics for evaluating the TT model with respect to the ground truth from the dataset. The results of these metrics should be comparable (e.g. within a couple percentage points of difference) to those obtained from running the evaluation with the reference model on CPU / GPU. The metrics used can include the [perplexity](https://en.wikipedia.org/wiki/Perplexity) (computed as $e^{\text{(avg negative log likelihood)}}$, i.e. how well the LLM predicts the text in the dataset), and the top-1/5 accuracy of the predicted next token.
+
+#### Debugging Accuracy
```

Comment:
Other possible workarounds/things to try out:
- Convert inputs and outputs to DRAM interleaved so that the problematic op reads/writes to DRAM
- Remove custom program configs and try the ttnn default
- For CCL: double checking reduction dims (mostly for 2D weight sharding)
- For MM and if anything changes around weight source, transposed weights etc: comment cache loading to make sure the weights are generated from the torch tensor and not loaded from disk

Reply:
Thanks @uaydonat @johanna-rock-tt, I have added these additional scenarios to the debugging section.

Reply:
A couple more since I requested other changes:
* make sure the sharding of tensors matches between the producer and consumer ops
* check any bug fixes to the reference models on Github or Huggingface (e.g. llama3.2 posted by meta had a bug initially, which was fixed sometime later).

Reply:
Added these additional debugging tips